dev/tools/gen_defaults/lib/input_decorator_template.dart

```diff
@@ -43,20 +43,20 @@ class _${blockName}DefaultsM3 extends InputDecorationTheme {
       return ${border('md.comp.filled-text-field.disabled.active-indicator')};
     }
     if (states.contains(MaterialState.error)) {
-      if (states.contains(MaterialState.hovered)) {
-        return ${border('md.comp.filled-text-field.error.hover.active-indicator')};
-      }
       if (states.contains(MaterialState.focused)) {
         return ${mergedBorder('md.comp.filled-text-field.error.focus.active-indicator','md.comp.filled-text-field.focus.active-indicator')};
       }
+      if (states.contains(MaterialState.hovered)) {
```

Comment:
Just curious why we have the change here. When it is focused, we still want to see the hover effect, right? So probably hovered state should be put above focused?

Reply:
This was a change from another PR which wasn't synced with the template class. I synced it here to pass the new check. 

cc: @bleroux

Reply:
Ah I see. I found the PR change: https://github.com/flutter/flutter/pull/146127
Looks like this is expected, can we add a comment above so this will not cause any confusion. Usually we'll have a order: pressed->hovered->focused. Here is a little more explanation: https://github.com/flutter/flutter/pull/125905 😄

Reply:
Hey @QuncCccccc!

My PR which introduced this change, https://github.com/flutter/flutter/pull/146127, landed just before Taha implemented this new analyser test. Thanks to this new test Taha spotted that I forgot to update the template. Instead of me filing a PR to update the template we decided that it will be ok to do it in this PR.

About the change itself, the problem with having the hovered state taking precedence over the focused one is especially noticeable on text field: the focused state for a text field increases border width (2dp) and applies bright colors (primary color or error color) while the hover state has the same border than a non focused text field (1dp) and use a color a little darker that a non focused text field. On desktop, it is also very common that a text field is focused and hovered (when users rely mainly on their mouse), so with the previous implementation when the text field was hovered it feels more or less as it lost focus:

https://github.com/flutter/flutter/assets/840911/a882e39e-8013-4854-bf4e-5ecf3cf3de0e

The M3 specification did not specify the exact behavior for this kind of visual properties impacted by hovered/focused states. What they explain in details is that hover is meant as a layer above the component but that is applicable for filled area not for text style and border like this.

I compared with other M3 implementations, especially https://material-web.dev/components/text-field/, and focus state takes precedence over hovered state for text field border (and also for the label color, that's not the case for Flutter text field for the moment).

I plan to file a similar PR for the label text color, I think it will be great to discuss that topic together on this coming PR, there might be cases I missed.

Reply:
Got it. This totally makes sense to me! Thanks for the explanation! I think adding a comment above would be helpful because other widgets have a reversed order:)

Reply:
> Got it. This totally makes sense to me! Thanks for the explanation! I think adding a comment above would be helpful because other widgets have a reversed order:)

To avoid blocking Taha's PR, I added the detailed comment on https://github.com/flutter/flutter/pull/146572